### ISSUE TITLE
Improve Performance of Goto Implementation

### ISSUE THREAD

**Issue by dantleech** [phpactor/phpactor#873]
Goto Implementation (find class implementations) is so slow - this is because by default it scans the whole vendor directory and needs to reflect each class to see if it, somehow, implements the target class.

There are some stategies put in place to pre-filter the set (e.g. path similarity) but these are unreliable.

This ticket is to track an investigation to see if the performance can be improved.

### Parsing bottleneck?

Blackfire shows there is a [parsing bottleneck](https://blackfire.io/profiles/cb2c6dcc-e0e6-4f70-bcad-aaeaa8f12981/graph?settings%5Bdimension%5D=wt&settings%5Bdisplay%5D=landscape&settings%5BtabPane%5D=nodes&selected=Microsoft%5CPhpParser%5CParser%3A%3AparseDelimitedList%3F1f50ef16e47610ddcd5ab1d3d0ebc34c&callname=main())

When trying to mitigate this with a filesystem cache (md5 / unserialize) we see an significant improvement [on Blackfire](https://blackfire.io/profiles/8c70cfa1-100c-4fce-b049-48bcb1b608d2/graph?settings%5Bdimension%5D=wt&settings%5Bdisplay%5D=landscape&settings%5BtabPane%5D=nodes&selected=&callname=main()).

Filesystem cache + in memory cache: `~853ms`
In memory cache: `1.3s`

However when running the operation locally with `time phpactor rpc` there is hardly any difference between the filesystem-cached version and the current version (which caches parsing results in-memory).

Filesystem cache + in memory cache: 
`./bin/phpactor rpc  0.38s user 0.06s system 100% cpu 0.435 total`
In memory cache:
`./bin/phpactor rpc  0.33s user 0.07s system 100% cpu 0.408 `

Perhpaps the Blackfire are skweed by the amount of operations performed when parsing so many files:?  

### Caching Worse Reflection

Using a worse-reflection level cache results in huge (sometimes > 4MB) cache files per class. The `unserialize` operation is then the [bottleneck](https://blackfire.io/profiles/dc9201cc-6e2f-40e6-911b-f1b421393f72/graph?settings%5Bdimension%5D=wt&settings%5Bdisplay%5D=landscape&settings%5BtabPane%5D=nodes&selected=&callname=main()). 

### Indexing

We could automatically index all classes reflected by Worse Reflection and
provide a query API (e.g. all classes implementing X)

The index can be self-invalidating, however the index will not be able to pick
up new files.

So we will need to perform a seach for files with a modified/created date greater
than the indexes last-updated timestamp.

**Comment by elythyr** [comment]
The indexing seems to be the better solution but it's also the most complicated one.
We first need to find an indexer (I don't think we should try to implement one).
Then we need to provide an endpoint to query it.
Finally we need a way to generate it in the background, with vim we could make the plugin start a job in the background.
Or just launch a script with some kind of inotify...

It's quite a change, I think at some point we won't have the choice in order to improve performances.
But we could start by using it only on some specific feature that requires it in order to test it.

I don't know anything about indexing so I won't grant a lot to the table, but it seems like an interesting topic if we decide to go there.

**Comment by dantleech** [comment]
I've actually got some way with this extension: https://github.com/phpactor/project-query

It builds a file-based index. The index can take minutes (scaling with the size of the project) to build but updating it will typically take less than a second (we only need to reindex those files whose timestamp is greater than the last update of the index).

I tested this and it works pretty well. 

We can "update" the index on each request and it will still be faster than the current goto implementatoin. The initial index build can take some time and I think this can be executed as a command external to f.e. VIM (or possibly asynchronously) as it's a one-time task per-project.

Still some minor things to clear up, but I think it can be in the release.
